### PR TITLE
paginated inbox

### PR DESCRIPTION
**Pull request by catdevnull**
fixes https://github.com/hyphacoop/distributed-press-organizing/issues/163

- [ ] add some way to automatically add indexes to existing inboxes

### CODE REVIEW

File: src/server/store/ActivityStore.ts [hunk 1]
```diff
@@ -20,11 +22,36 @@ export class ActivityStore {
...
+      .sublevel('indexes', { valueEncoding: 'json' })
+  }
+
+  get publishedIndex (): AbstractLevel<any, string, any> {
+    return this.indexesDB
+      .sublevel('published', { valueEncoding: 'json' })
+  }
+
+  async addToIndex (publishedAt: Date, id: string): Promise<void> {
+    await this.publishedIndex.put(`${publishedAt.toISOString()}-${id}`, id)
```
[the remark this answers is not in the document]
Comment (catdevnull) [reply src/server/store/ActivityStore.ts:45]: I'm believe toISOString produces sortable strings:

>The toISOString() method of Date instances returns a string representing this date in the date time string format, a simplified format based on ISO 8601

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString

>Date and time values are ordered from the largest to smallest unit of time: year, month (or week), day, hour, minute, second, and fraction of second. The lexicographical order of the representation thus corresponds to chronological order

https://en.wikipedia.org/wiki/ISO_8601

File: src/server/store/ActivityStore.ts [hunk 2]
```diff
@@ -37,11 +64,52 @@ export class ActivityStore {
...
+        continue
+      }
+      activities.push(await this.db.get(id))
     }
     return activities
   }
+
+  async count (): Promise<number> {
+    let count = 0
+    for await (const _ of this.db.keys()) { // eslint-disable-line @typescript-eslint/no-unused-vars
```
[the remark this answers is not in the document]
Comment (catdevnull) [reply src/server/store/ActivityStore.ts:103]: I'm not sure how the test would be implemented. As I understand it, this wouldn't be possible, as sublevels are entirely different databases, right?

[the remark this answers is not in the document]
Comment (RangerMauve) [reply src/server/store/ActivityStore.ts:109]: Just test that after inserting three documents the count is three.

[the remark this answers is not in the document]
Comment (RangerMauve) [reply src/server/store/ActivityStore.ts:109]: Sublevels are key prexfixes over the parent database.

e.g. `sub(db, "example")` would prefix all keys with `example`. that way you can more easily partition the key space.

File: src/server/store/ActivityStore.ts [hunk 3]
```diff
@@ -22,13 +23,13 @@ export class ActivityStore {
...
     await this.db.put(key, activity)
-    await this.addToIndex(getPublished(activity.published), key)
+    await this.addToIndex(activity)
   }
 
   async remove (url: string): Promise<void> {
     const key = this.urlToKey(url)
     await this.db.del(key)
-    await this.removeFromIndex(key)
+    await this.removeFromIndex(url)
```
Comment (RangerMauve) [inline src/server/store/ActivityStore.ts:32]: We should remove from the index before the del, else it won't be able to fetch the doc.

Mind adding a test for this? With a fresh store: Add a doc, remove the doc by url, test that there are no index keys and no docs.

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Great! Lets just get some small tests in to make sure it's working and then it looks good to me. :)

File: src/server/store/ActivityStore.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (RangerMauve) [reply src/server/store/ActivityStore.ts:110]: This was the case, only affected the migration script. Fixed.

File: src/server/store/ActivityStore.ts (hunk 3 above)
Comment (RangerMauve) [reply src/server/store/ActivityStore.ts:31]: Fixed tests, modified list test to check ordering.

**Pull request merged**